src/openpa/src/primitives/opa_gcc_intrinsics.h

```diff
@@ -34,54 +42,86 @@ typedef struct {
    may not be true at all. */
 static inline int OPA_load_int(const OPA_int_t * ptr)
 {
+#ifdef USE_GCC_ATOMIC
+    return __atomic_load_n(&ptr->v, __ATOMIC_RELAXED);
+#else
     return ptr->v;
+#endif
 }
 
 static inline void OPA_store_int(OPA_int_t * ptr, int val)
 {
+#ifdef USE_GCC_ATOMIC
+    __atomic_store_n(&ptr->v, val, __ATOMIC_RELAXED);
+#else
     ptr->v = val;
+#endif
 }
 
 static inline void *OPA_load_ptr(const OPA_ptr_t * ptr)
 {
+#ifdef USE_GCC_ATOMIC
+    return __atomic_load_n(&ptr->v, __ATOMIC_RELAXED);
+#else
     return ptr->v;
+#endif
 }
 
 static inline void OPA_store_ptr(OPA_ptr_t * ptr, void *val)
 {
+#ifdef USE_GCC_ATOMIC
+    __atomic_store_n(&ptr->v, val, __ATOMIC_RELAXED);
+#else
     ptr->v = val;
+#endif
 }
 
 static inline int OPA_load_acquire_int(const OPA_int_t * ptr)
 {
+#ifdef USE_GCC_ATOMIC
+    return __atomic_load_n(&ptr->v, __ATOMIC_ACQUIRE);
+#else
     volatile int i = 0;
     int tmp;
-    tmp = ptr->v;
     __sync_lock_test_and_set(&i, 1);    /* guarantees acquire semantics */
+    tmp = ptr->v;
```

Comment:
I'm not sure if this reordering is correct.
Consider the following:
```
int val1 = OPA_load_acquire_int(&atomic_val);
int val2 = *memory;
```
The acquire semantics guarantees that this `atomic_val` load should be done *before* `int val2 = *memory;`
However, your current code would look like the following after inlining:
```
__sync_lock_test_and_set(...);
int val1 = atomic_val->v;
int val2 = *memory;
```
Obviously, `atomic_val` load and `memory` load can be reversed, which breaks the acquire ordering.

Of course, this `__sync`-based implementation is not complete. The original one lacks a kind of memory flush operation that (somehow) guarantees this thread read a value written by other threads. Do we need two?

Reply:
Thinking on this part a bit more, my implementation is wrong, as was the original.  But even with 2 load-acquires, this will not work because of the semantics of load acquire allow older loads in program order to order after the barrier and mix with younger memory instructions in program order you are trying to order with.  So:

```
__sync_lock_test_and_set(&i, 1);
tmp = ptr->v;
__sync_lock_test_and_set(&i, 2);
<memory to be order after tmp reads ptr->v>
```
The tmp variable could still be ordered after the younger reads because it is older than the second acquire, and we end up with a race. 

In this case, what is actually needed is a full barrier to guarantee the ordering of the first read with younger reads because the `__sync` primitives lack real acquire semantics. In the case of Armv8 asm, a `DMB LD` would suffice.

Ref: page B2-95 of the Armv8 Architecture Reference Manual.

Reply:
```
// OPA_load_acquire_int starts.
__sync_lock_test_and_set(&i, 1);
tmp = ptr->v;
__sync_lock_test_and_set(&i, 2);
// OPA_load_acquire_int ends.
int val = *mem;
```

- `load(mem)` must be ordered after `load(ptr->v)`.
Yes, you're right. That is what `OPA_load_acquire_int()` needs to guarantee in terms of acquire semantics.

- In this implementation, `load(ptr->v)` could be executed after `load(mem)`.
I don't think so.

Your reason seems because
> the semantics of load acquire allow older loads in program order to order after the barrier and mix with younger memory instructions in program order you are trying to order with.

I don't think the load acquire allows older loads in program order to be executed before the acquire load.

- [Reason 1] Spec.
https://gcc.gnu.org/onlinedocs/gcc-4.1.0/gcc/Atomic-Builtins.html
```
__sync_lock_test_and_set:
This means that references after the builtin cannot move to (or be speculated to) before the builtin,
but previous memory stores may not be globally visible yet, and previous memory loads
may not yet be satisfied.
```
So, references after the builtin (i.e., older loads) cannot move to before the builtin (i.e., `__sync_lock_test_and_set(&i, 2)`).

- [Reason 2] Intuition
```
int value = 0;
void thread0(int *p_done) {
  while (1) {
    if (acquire_load(p_done) == 1) { // ex1-l1
       printf("%d\n", value);        // ex1-l2
       return;
    }
}
// Thread 1
void thread0(int *p_done) {
  value = 1;
  acquire_store(p_done, 1);
}
```
Intuitively, `ex1-l2` (i.e., an older load) should be executed after the load at `ex1-l1` (i.e., younger acquire load). Otherwise, the acquire load is useless.

Reply:
Note: why and when might `__sync_lock_test_and_set(&i, 1);` be useful?
In theory, without acq/rel/seq memory operations, the program can always read a value that is (something like) a locally cached value.
```
int value = 0;
void thread0(int *p_done) {
  sleep(1000);
  int done = *p_done; // load a locally cached *p_done.
  _sync_lock_test_and_set(&i, 2);
  // done can be always 0.
}
// Thread 1
void thread0(int *p_done) {
  acquire_store(p_done) == 1;  
}
```
In practice, load is repeated (and the current load implementation contains a acquire barrier), so load is not a big problem.

This issue might be more significant in the case of store.
```
int value = 0;
void thread0(int *p_done) {
  int done = acquire_load(p_done);
  // done can be always 0.
}
// Thread 1
void thread0(int *p_done) {
  __sync_lock_release(...);
  acquire_store(p_done, 1); // it writes a value to a kind of a local cache
                            // so this write is never globally visible.
}
```

In reality, this kind of cache is somehow flushed eventually, so there's no problem, but semantically it is not guaranteed.

Reply:
Please point out if my understanding is wrong.

Reply:
I believe we're crossing terminology a bit here, my definition for an _older_ load is one that appears earlier in the instruction stream of the micro-processor (so older is not always corresponding to a lower line number in the program, in case we're in a loop and speculating).  

Load-acquire does indeed allow **older** loads to be observed after the **younger** barrier is observed.  It is explicitly stated in the GCC definition:

Adding under-scores for emphasis.
```
__sync_lock_test_and_set:
This means that references __after__ the builtin cannot move to (or be speculated to) before the builtin,
but __previous__ memory stores may not be globally visible yet, and __previous__ memory loads
may not yet be satisfied.
```
The spec does specify that _previous_ (read _older_) loads and stores may not have fully executed yet when the barrier completes, but references _after_ (read _younger_) must wait for the barrier.  So older references are effectively unordered with respect to the acquire barrier.

In your first example:
> ```
> int value = 0;
>void thread0(int *p_done) {
>  while (1) {
>    if (acquire_load(p_done) == 1) { // ex1-l1
>       printf("%d\n", value);        // ex1-l2
>       return;
>    }
>}
>// Thread 1
>void thread1(int *p_done) {
>  value = 1;
>  acquire_store(p_done, 1);
>}
>```
Thread0 is correctly using an acquire barrier, the loop will execute a bunch of `load-acquire p_done` and will have to delay the read of `value`. When processor does eventually observe `p_done == 1`, it will not have been allowed to speculatively load `value` ahead of seeing `p_done == 1`.  Thread1 using an acquire for the store is incorrect (I figure this is a mistake, and you meant release, but lets assume we had an acquire here), because the acquire would allow `value = 1` to be executed **after** `p_done = 1`.  Therefore Thread0 could see `p_done == 1` and `value == 0` when it is intended that Thread0 should only see `value == 1` when `p_done` is set.  That store would have to be a _release_ to properly order `value` being set and observed before `p_done`.

>In reality, this kind of cache is somehow flushed eventually, so there's no problem, but semantically it is not guaranteed.

Flushing memory is not what barriers do, they explicitly enforce an order on the memory reference stream as observed from a remote entity in the system.  It is entirely up to the micro-architects to design how the barriers enforce that ordering and where.  In some systems, a barrier was actually emitted as a transaction on the memory fabric to order reads/writes and not something just executed by the core.

Reply:
Thanks for your correction. You're right.

- I was confused by older/younger usage.
- `acquire_store` should be `release_store` in the figure. There's no `acquire_store`. Thank you for your correction.
- Flushing memory: You're right, a barrier is not a "flush" operation.

### 1. Need of a barrier (or some synchronization mechanism) after store.

Strictly speaking, I think a barrier is needed after store.
For example, a normal write does not guarantee its visibility to other threads.

```
// C++ 11 draft (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf), S.1.10.25,
An implementation should ensure that the last value (in modification order) assigned by an atomic or
synchronization operation will become visible to all other threads in a finite period of time.
```

Normal load/stores are neither atomic nor synchronization operations.
So theoretically, in my understanding, `val = 1` might not be visible to `thread1`.

```c
int val = 0;
void thread0() {
  // OPA_store_release_int starts
  [barrier]; // either __sync_lock_release or __sync_synchronize
  val = 1;
  //    <= Point A
  // OPA_store_release_int ends
  while (1);
}
void thread1() {
  while (OPA_load_acquire_int(&val) == 0);
  kill_thread(thread0_handle);
}
```

We might need an atomic/synchronization operation (e.g., a release barrier) at `Point A`.
In addition, if the `relaxed` one (i.e., `OPA_store`) should follow the atomic operation semantics (since a `relaxed` operation is atomic), we might need to put a release barrier after `OPA_store`.

### 2. Semantics of `OPA_load_acquire`

I'd like to clarify the semantics you assume.

```c
int val1 = *mem1;
int val2 = OPA_load_acquire_int(mem2);
int val3 = *mem3;
```

`OPA_load_acquire` must guarantee the following:
- Younger loads (i.e., `*mem3`) need to be executed after load-acquire (i.e, `OPA_load_acquire_int(mem2)`).

This one should be satisfied by putting  `__sync_lock_test_and_set()`.

In my understanding, `OPA_load_acquire` does NOT require the following because its name is `load_acquire`:
- Older loads (i.e., `*mem1`) need to be executed before load-acquire (i.e, `OPA_load_acquire_int(mem2)`).

Do you think we need to ensure this as well?

Reply:
In regards to your first comment:

> Normal load/stores are neither atomic nor synchronization operations.
> So theoretically, in my understanding, val = 1 might not be visible to thread1.
> ```C
>int val = 0;
>void thread0() {
>  // OPA_store_release_int starts
>  [barrier]; // either __sync_lock_release or __sync_synchronize
>  val = 1;
>  //    <= Point A
>  // OPA_store_release_int ends
>  while (1);
>}
>void thread1() {
>  while (OPA_load_acquire_int(&val) == 0);
>  kill_thread(thread0_handle);
>}
>```

I have not studied the C/C++ memory model in detail (so far I've been referencing a CPU architecture memory model), but my understanding is that for coherent memory systems, the above code fragment will work as intended. `int val` is defined as a global and is used in two functions, so the compiler will not be able to optimize the read and writes to `val` to be in CPU local registers.  The hardware will then take care of propagating the write in thread0 to thread1, and generally an int can be written atomically by modern systems.  Since we're dealing with one location in memory, no barriers would be required either, a relaxed ordered would be fine.  The issue here is it appears the C/C++ memory model is assuming an abstract machine that may not guarantee loads/stores to ints as being atomic.  Maybe the machine you're running on only loads and stores a bit at a time, in that case for the above code to be portable, you would need to wrap the write to val in thread0 with `__atomic_store_n()` and the read in thread1 with `__atomic_load_n()`. The variant of OPA using the `__atomic` builtin's would be guaranteed to be portable.

In regards to your second:

>```C
>int val1 = *mem1;
>int val2 = OPA_load_acquire_int(mem2);
>int val3 = *mem3;
>```

The proper ordering for the load acquire are:
1. Reference to `mem2` is ordered and visible before the reference to `mem3`.
1. Reference to `mem1` is **not** ordered before any reference.

In regards to the `__sync_lock_test_and_set()` is that the current implementation of OPA, it uses a dummy value to operate on because `__sync_lock_test_and_set()` is not guaranteed to be able to write anything to the passed in pointer but 1.  Therefore the fragment expands to:

```C
int val1 = *mem1;
int val2 = *mem2;
int dummy = __sync_lock_test_and_set(&dummy, 1);
int val3 = *mem3;
```

Because we have now inside the function introduced a new variable, the acquire semantics of `__sync_lock_test_and_set()` only guarantee the following:
1. Reference to `dummy` is ordered and visible before the reference to `mem3`
1. Reference to `mem1` is **not** ordered
1. Reference to `mem2` in **not** ordered

This no longer follows the assumed behavior of a `load_acquire(mem2)`.  It is the introduction of the dummy value that makes the original implementation of `OPA_load_acquire` incorrect, and by extension `OPA_store_release` as well.  The only way to get the proper behavior with `__sync` is to remove `__sync_lock_test_and_set()` and replace it with the stronger `__sync_synchronize()`.  

In the long run, the best solution for MPICH would be to drop support for older GCC's when  able and switch to just the `__atomic` builtins instead of carrying the multiple implementations.

Reply:
### Need of a barrier (or some synchronization mechanism) after store.

It's not about the implementation of the hardware, but more about the semantics of C/C++ programming language. For example, consider the following weird program:

```c
int atomic_val = 0;
void thread(int busy_loop) {
  // OPA_store_release_int starts
  __sync_synchronize();
  atomic_val = 1;
  // OPA_store_release_int ends

  while (busy_loop);

  // OPA_store_int starts
  atomic_val = 2;
  // OPA_store_int ends
}
void thread2() {
  while (OPA_load_acquire_int(&atomic_val) == 0);
  kill_thread(thread_handle);
}
```

In release-store semantics, even if `busy_loop` is 1, `atomic_val` (or `OPA_store_release_int`) must be written to the register. However, this `atomic_val = 1;` in `OPA_store_release_int` is neither an atomic or synchronization operation, so the compiler can eliminate this write as follows.

```
// gcc 5.4.0 on Ubuntu 16.04, Intel x86/64
// gcc cmp.c -O3 -S
        .globl  thread
        .type   thread, @function
thread:
.LFB0:
        .cfi_startproc
        testl   %edi, %edi
        mfence
        je      .L2
.L4:
        jmp     .L4
        .p2align 4,,10
        .p2align 3
.L2:
        movl    $2, atomic_val(%rip)
        ret
```
Basically, if `je .L2` holds (the first argument `busy_loop` (=`%edi`) is non-zero), it will enter a busy loop (`jmp .L4`), but before this loop there's no `atomic_val = 1`.

In any case, the program should work as specified by the programming language, and under some circumstances (compilers and underlying hardware), `OPA_store_release_int ` without a barrier can be broken.

The question is whether such a case happens in reality or not. `__sync_synchronize()` is heavy, so we should avoid it if possible.

### Use __sync_synchronize

On the second thought, you're right. We should use `__sync_synchronize`. Thanks for your correction!

---

I agree that we should drop `__sync` and move to c11 atomic or `__atomic` in the future.

About performance, the updated `__sync` version is correct but seems more heavyweight. We should use `__atomic` if possible to minimize the performance loss, but I am not sure if the current detection (`__GNUC__`, `__GNUC_MINOR__`, and `__GNUC_PATCHLEVEL__` > 4.7.4) is sufficient. I am not sure if it should be in this PR, but we should at least examine if it works with other compilers (e.g., clang, icc, ...).

Reply:
I now understand what you mean in terms of the C/C++ memory semantics.  The compiler is entirely within spec to eliminate what it identifies as a redundant store.  I agree, `__sync_synchronize()` should be used sparingly.  I've done a few experiments to mimic what OPA is doing by modifying your example to be the following:

```C
struct atomic {
  volatile int val;
};
struct atomic atomic_val = {0};

void atomic_store(struct atomic *val, int v) {
  val->val = v;
}

void thread(int busy_loop) {
  atomic_store(&atomic_val, 1);
  while(busy_loop);
  atomic_store(&atomic_val, 2);
}
```

Because the variable inside OPA's definition of an atomic is volatile, the compiler is disallowed from doing dead code analysis and removing that initial write that could break the program.  In this case, the `__sync` version of OPA should be ok still.  Using the newer `__atomic` interface does remove the necessity of creating the above construct to prevent the compiler from re-arranging and omitting memory references itself.  

## Is the compiler detection macro enough?

I haven't tested on clang or any proprietary compilers like ICC.  I would suspect that it would not work.  But I am not familiar enough with how MPICH configures its Makefiles with different compilers, would the `opa_gcc_intrinsics.h` header even be selected using a different compiler?  I think that should be a different PR to make changes to detecting non-GCC compilers.

Reply:
### Implementation of relaxed store with __sync primitives
I agree that, with `volatile`, this kind of problem seems less likely to happen though it is not 100% guaranteed by the language specification; the language does not guarantee the volatile memory write will be visible to other threads.
```
// C++ 11 draft (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf), S.1.10.25,
An implementation should ensure that the last value (in modification order) assigned by an atomic or synchronization operation will become visible to all other threads in a finite period of time.
```

However, at least I don't know (and can't imagine) the combination of compilers and architectures that can cause this issue in practice. In `OPA_store_int ` (relaxed one), we don't need `__sync_synchronize()`.

### Compiler macro
icc and clang (and some other compilers (one for Solaris)) use `opa_gcc_intrinsics.h` as far as I remember. We might need a different PR for better compiler detection.

Reply:
A different PR is probably warranted for the new support.  I can try on clang though to make sure it compiles at least.  I do not have access to any Solaris machines, or ICC.

Reply:
Created #4231.
That PR is simple enough, so you might want to use it in this PR; if so please cherry-pick #4231.